File: bindings/python/m9m/engine.py
"""WorkflowEngine wrapper for the m9m Python SDK."""

from __future__ import annotations

import json as _json
import subprocess
import tempfile
from typing import Any

from .binary import get_binary_path
from .data import DataItem, ExecutionResult
from .workflow import Workflow


class WorkflowEngine:
    """Spawns the m9m binary to execute workflows."""

    def __init__(self, credential_manager: Any | None = None):
        self._binary = get_binary_path()
        self._credential_manager = credential_manager
        self._custom_nodes: dict[str, Any] = {}

    def execute(
        self,
        workflow: Workflow,
        input_data: list[dict[str, Any]] | None = None,
    ) -> ExecutionResult:
        """Execute a workflow with optional input data."""
        # Write workflow to a temp file
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            f.write(workflow.to_json())
            workflow_path = f.name

        input_data = input_data or []
        input_json = _json.dumps(input_data)

        try:
            result = subprocess.run(
                [self._binary, "exec", workflow_path, "--input", input_json],
                capture_output=True,
                text=True,
                check=False,
            )
            if result.returncode != 0:
                return ExecutionResult(error=result.stderr.strip() or "execution failed")
            # Try to parse JSON output
            try:
                return ExecutionResult.from_json(result.stdout)
            except Exception:
                # If output is not JSON, wrap it
                return ExecutionResult(
                    data=[DataItem(json={"output": result.stdout})]
                )
        except FileNotFoundError:
            return ExecutionResult(error="m9m binary not found")
        finally:
            import os

            try:
                os.unlink(workflow_path)
            except OSError:
                pass

    def register_node(self, name: str, handler: Any) -> None:
        """Register a custom node handler (placeholder — requires server mode)."""
        self._custom_nodes[name] = handler

    def node(self, name: str, **kwargs: Any) -> Any:
        """Decorator for registering custom nodes."""

        def decorator(func: Any) -> Any:
            self.register_node(name, func)
            return func

        return decorator

    def __enter__(self) -> "WorkflowEngine":
        return self

    def __exit__(self, *args: Any) -> None:
        pass

File: packages/m9m-cli/nodejs/src/engine.ts
/**
 * WorkflowEngine class for executing workflows via the m9m binary.
 */

import { spawn } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { getBinaryPath, downloadBinary } from './binary';
import { Workflow } from './workflow';
import type { DataItem, ExecutionResult, NodeExecutorFn } from './types';

function runBinary(args: string[], stdin?: string): Promise<{ stdout: string; stderr: string; code: number }> {
  return new Promise((resolve, reject) => {
    const binary = getBinaryPath();
    const child = spawn(binary, args, { stdio: ['pipe', 'pipe', 'pipe'] });

    let stdout = '';
    let stderr = '';

    child.stdout.on('data', (data) => (stdout += data));
    child.stderr.on('data', (data) => (stderr += data));

    child.on('close', (code) => {
      resolve({ stdout, stderr, code: code ?? 1 });
    });

    child.on('error', reject);

    if (stdin) {
      child.stdin.write(stdin);
    }
    child.stdin.end();
  });
}

/**
 * High-level interface to the m9m workflow execution engine.
 *
 * @example
 * ```typescript
 * const engine = new WorkflowEngine();
 *
 * // Load and execute a workflow
 * const workflow = Workflow.fromFile('workflow.json');
 * const result = await engine.execute(workflow);
 * ```
 */
export class WorkflowEngine {
  private customNodes: Map<string, NodeExecutorFn> = new Map();

  /**
   * Create a new workflow engine.
   */
  constructor(_options?: { credentialManager?: unknown }) {
    // Binary is auto-downloaded on first use if needed
  }

  /**
   * Ensure the binary is available (downloads if needed).
   */
  async ensureBinary(): Promise<void> {
    try {
      getBinaryPath();
    } catch {
      await downloadBinary();
    }
  }

  /**
   * Execute a workflow with optional input data.
   */
  async execute(workflow: Workflow, inputData?: DataItem[]): Promise<ExecutionResult> {
    await this.ensureBinary();

    // Write workflow to temp file
    const tmpDir = os.tmpdir();
    const workflowPath = path.join(tmpDir, `m9m-workflow-${Date.now()}.json`);
    fs.writeFileSync(workflowPath, JSON.stringify(workflow.toJSON()));

    try {
      const inputJson = inputData ? JSON.stringify(inputData) : '[]';
      const { stdout, stderr, code } = await runBinary(
        ['exec', workflowPath, '--input', inputJson]
      );

      if (code !== 0) {
        return { data: [], error: stderr || 'execution failed' };
      }

      try {
        const parsed = JSON.parse(stdout);
        return { data: parsed.data || [], error: parsed.error };
      } catch {
        return { data: [{ json: { output: stdout } }], error: undefined };
      }
    } finally {
      try {
        fs.unlinkSync(workflowPath);
      } catch {
        // ignore
      }
    }
  }

  /**
   * Load a workflow from a JSON file.
   */
  loadWorkflow(path: string): Workflow {
    return Workflow.fromFile(path);
  }

  /**
   * Parse a workflow from a JSON string or object.
   */
  parseWorkflow(json: string | object): Workflow {
    return Workflow.fromJSON(json);
  }

  /**
   * Register a custom node type.
   */
  registerNode(nodeType: string, executor: NodeExecutorFn): void {
    this.customNodes.set(nodeType, executor);
  }

  /**
   * Decorator-style method to register a node.
   */
  node(nodeType: string): (executor: NodeExecutorFn) => NodeExecutorFn {
    return (executor: NodeExecutorFn): NodeExecutorFn => {
      this.registerNode(nodeType, executor);
      return executor;
    };
  }
}
